File: 435.non-overlapping-intervals.py
# @leet start
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        result = 0
        intervals.sort(key=lambda x: (x[0], abs(x[1]-x[0])))
        print(intervals)

        left, right = intervals[0][0], intervals[0][1]

        for i in range(1, len(intervals)):
            v_left, v_right = intervals[i][0], intervals[i][1]
            start = max(left, v_left)
            end = min(right, v_right)

            if start < end:
                result += 1
                if right - left > v_right - v_left:
                    left, right = v_left, v_right
            else:
                left, right = v_left, v_right

        return result
# @leet end

File: 435.non-overlapping-intervals.cpp
// @leet start
bool compare(const vector<int> &a, vector<int> &b) {
    if (a[1] != b[1]) {
        return a[1] < b[1];
    } else {
        return a[0] < b[0];
    }
}

class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        int result = 0;
        sort(intervals.begin(), intervals.end(), compare);

        int int_right = intervals[0][1];

        for (int i = 1; i < intervals.size(); i++) {
            int c_right = intervals[i][1];
            int c_left = intervals[i][0];

            if (int_right > c_left) {
                result++;
            } else {
                int_right = c_right;
            }


        }

        return result;
    }
};
// @leet end
